neps/nep-0413.md

```diff
@@ -0,0 +1,135 @@
+The most common flow for [NEAR user authentication into a Web3 frontend](https://docs.near.org/develop/integrate/frontend#user-sign-in--sign-out) involves the creation of a [FunctionCall Key](](https://docs.near.org/concepts/basics/accounts/access-keys)).
+
+One might feel tempted to reproduce such process here, for example, by creating a key that can only be used to call a non-existing method in the user's account. This is a bad idea for several reasons:
+1. It implies that the user needs to expend gas in creating a new key.
+2. Since any third-party can ask the user to create a `FunctionCall Key`, this opens a vector of attack.
+
+Using a FullAccess key allows us to be sure that the message was signed by the user (since nobody should have access to their `FullAccess Key`), while keeping the constraints of not expending gas in the process (because no new key needs to be created).
+
+### How to Return the Signed Message in a Safe Way
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since the fragment isn't included in the Referrer.
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Given the previous [rationales](#rationale-and-alternatives), this NEP proposes the wallets to implement a `verifyOwner` method following the specifications bellow.
+
+### Goal
+The `verifyOwner` method must take as input a `domain` and a `message`, and transform them into a signature.
+
+### Input Interface
+In order to sign a message, `verifyOwner` must implement the following input interface:
+
+```jsx
+interface VerifyOwnerParams {
+  domain: string; // The app in which the user wants to authenticate, e.g. myapp.com.
+  message: string; // A message to be included in the signature
+  callbackUrl?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). This is the callback url once the signing is approved/cancelled. Defaults to `window.location.href`.
+}
+```
+
+### Structure and Signature
+In order to respect all the points layed in [rationales](#rationale-and-alternatives), `verifyOwner` must:
+
+1. Embed the passed `domain` and `message` into a predefined structure (see below).
+2. Stringify the structure, respecting the attribute's order (see bellow).
+2. Compute the `SHA256 hash` of the *stringified structure*.
+3. Use a full-access key to sign the resulting `SHA256 hash`.
+
+#### Structure
+The structure must have the following interface:
+
+```rust
+struct Payload {
+  accountId: string; // The account name as plain text (e.g. "alice.near")
```

Comment:
Why do we need the accountId and the public key? Surely just the public key is sufficient?
```suggestion
```

Reply:
Indeed, since the public key necessarily belongs to an account, signing a message with the accountId included is redundant.